docs/Target Framework Strategy.md

```diff
@@ -0,0 +1,79 @@
+Target Framework Strategy
+===
+
+# Layers
+The roslyn repository produces components for a number of different products that push varying ship and TFM constraints on us. A summary of some of our dependencies are : 
+
+- Build Tools: requires us to ship compilers on `net472`
+- .NET SDK: requires us to ship compilers on current servicing target framework (presently `net6.0`)
+- Source build: requires us to `$(NetCurrent)` and `$(NetPrevious)` in workspaces and below (presently `net7.0` and `net6.0` respectively)
+- Visual Studio: requires us to ship `net472` for base IDE components and `net6.0` for private runtime components.
+
+It is not reasonable for us to take the union of all TFM and multi-target every single project to them. That would add several hundred compilations to any build operation which would in turn negatively impact our developer throughput. Instead we attempt to use the TFM where needed. That keeps our builds smaller but increases complexity a bit as we end up shipping a mix of TFM for binaries across our layers.
+
+# Require consistent API across Target Frameworks
+It is important that our shipping APIs maintain consistent API surface area across target frameworks. That is true whether the API is `public` or `internal`.
+
+The reason for `public` is standard design pattern. The reason for `internal` is a combination of the following problems:
+
+- Our repository make use of `InternalsVisibleTo` which effectively promotes `internal` APIs to `public` ones.
+- Our repository ships a mix of target frameworks. Typically workspaces and below will ship more recent TFMs than the layers above it. Compiler has to ship newer TFM for source build while IDE is constrained by Visual Studio's private runtime hence adopts newer TFM slower.
+- Our repository invests in polyfill APIs to make compiling against multiple TFMs in the same project a seamless experience.
+
+Taken together though this means that our `internal` surface area in many cases effectively `public` when it comes to binary compatibility. For example a consuming project can end up with `net7.0` binaries from workspaces layer and `net6.0` binaries from IDE layer. Because there is `InternalsVisibleTo` between these binaries the `internal` API surface area is effectively `public`. This requires us to have a consistent strategy for achieving 
+
+Consider a specific example of what goes wrong when our `internal` APIs are not consistent across TFM:
+
+- Workspaces today targets `net6.0` and `net7.0` and it contains our `EnumerableExtensions.cs` which polyfills many extensions methods on `IEnumerable<T>`. In `net7.0` the `OrderBy` extension method is not needed because it was put into the .NET core libraries.
```

Comment:
Nit: 

```suggestion
- Workspaces today targets `net6.0` and `net7.0` and it contains our `EnumerableExtensions.cs` which polyfills many extensions methods on `IEnumerable<T>`. In `net7.0` the `Order` extension method is not needed because it was put into the .NET core libraries.
```